### ISSUE TITLE
[GRM] Sort and apply resources sequentially

### ISSUE THREAD

**Issue by rfranzke** [gardener/gardener#5585]
**How to categorize this issue?**

/area usability
/kind enhancement

**What would you like to be added**:
Today, the `ManagedResource` controller part of `gardener-resource-manager` applies all resources in parallel. This deviates from what the [`chartrenderer`](https://github.com/gardener/gardener/blob/master/pkg/chartrenderer) and also upstream tools like `helm` or `kubectl` are doing.

The `ManagedResource` controllers should also [sort](https://github.com/gardener/gardener/blob/master/pkg/chartrenderer/sorter.go) the objects and then apply them (objects of the same kind could still be applied in parallel if desired).

**Why is this needed**:
We have seen issues where some resources were applied before others (e.g., `Deployment`s applied before `ServiceAccount`s) although it should rather be the other way round. For example, if a `ServiceAccount` specifies `automountServiceAccountToken=false` then the `Pod`s should not get a token volume, however, in rare cases it such wrong order can lead to such undesired situations.

**Comment by rfranzke** [comment]
It probably makes sense to tackle this after #4805 is closed.

**Comment by rfranzke** [comment]
FYI: #4805 is closed now, so work could be started if there is capacity.

**Comment by rfranzke** [comment]
/reopen
The PR https://github.com/gardener/gardener/pull/5822 gets reverted with #5898 due to https://github.com/gardener/gardener/issues/5893, @acumino can you check what is missing and reopen it?

**Comment by acumino** [comment]
/cc @rfranzke @ialidzhikov @timebertt If this is hard to follow we can discuss it in internal meeting.

I have created a new list in which resources need to be sorted before getting applied by the resource manager based on all resources that are available in Gardener.

**Sorted List Which we will use-**

CustomResourceDefinition
Namespace
HorizontalPodAutoscaler
Hvpa
VerticalPodAutoscalerCheckpoint
VerticalPodAutoscaler
ResourceQuota
LimitRange
PodSecurityPolicy
PodDisruptionBudget
Secret
SecretBinding
ConfigMap
StorageClass
PersistentVolume
PersistentVolumeClaim
ServiceAccount
CustomResourceDefinition
ClusterRole
ClusterRoleBinding
Role
RoleBinding
Service
PriorityClass
PodTemplate
DaemonSet
Pod
ReplicationController
ReplicaSet
Deployment
StatefulSet
Lease
Job
CronJob
IngressClass
Ingress
NetworkPolicy
APIService
CertificateSigningRequest
MutatingWebhookConfiguration
ValidatingWebhookConfiguration


**Resources which are not needed to be added in the list as they are not related to Manged resources** - 

ClusterOpenIDConnectPreset
OpenIDConnectPreset
CSIDriver
CSINode
App
DataObject
DeployItem
Execution
Installation
Target
BillingPeriod
MeteringReport
OpenIDConnect
TokenReview
LocalSubjectAccessReview
SelfSubjectAccessReview
SelfSubjectRulesReview
SubjectAccessReview
CloudProfile
ControllerDeployment
ControllerInstallation
ControllerRegistration
Plant
Project
Seed
Shoot
ShootState
Terminal
AuthCode
AuthRequest
Connector
DeviceRequest
DeviceToken
OAuth2Client
OfflineSessions
Password
RefreshToken
FlowSchema
ClusterBom
ClusterBomSync
ManagedResource
ManagedSeed
ManagedSeedSet
DNSEntry
DNSOwner
DNSProvider
Binding

**I am not sure about the order of these or do we even need them to added in the list**

ComponentStatus
Endpoints
Event
Node
ControllerRevision
BackupBucket
BackupEntry
ExposureClass
Quota
SigningKey
EndpointSlice
Event
BackupBucket
BackupEntry
Cluster
ContainerRuntime
ControlPlane
DNSRecord
Extension
Infrastructure
Network
OperatingSystemConfig
Worker
PriorityLevelConfiguration
RuntimeClass
Bastion

**Comment by ialidzhikov** [comment]
When I was debugging https://github.com/gardener/gardener/issues/5893 the problem was rather that in 

https://github.com/gardener/gardener/blob/0476c3f4c0fa57e3edbaf8b1f79247342551ff89/pkg/resourcemanager/controller/managedresource/sort.go#L91-L94

`a.oldInformation.Kind` and `b.oldInformation.Kind` were always empty strings and it was always comparing the objects by name.

**Comment by acumino** [comment]
> When I was debugging #5893 the problem was rather that in
> 
> https://github.com/gardener/gardener/blob/0476c3f4c0fa57e3edbaf8b1f79247342551ff89/pkg/resourcemanager/controller/managedresource/sort.go#L91-L94
> 
> `a.oldInformation.Kind` and `b.oldInformation.Kind` were always empty strings and it was always comparing the objects by name.

I checked once again it's not always comparing by name. It first tries to sort the object based on kind but if kind is not there in the list which we provide then it sorts based on the string comparison of a kind and if kind is the same then it sorts based on the object name.
Previously, VPA and others were not so they were getting deployed in last.

**Comment by acumino** [comment]
Also, I don't understand why the error message shows `autoscaling.k8s.io/v1beta2` haven't we already moved to `autoscaling.k8s.io/v1`. /cc @rfranzke 
I tried to reproduce the issue https://github.com/gardener/gardener/issues/5893. The crds were there with apiVersion `autoscaling.k8s.io/v1`. Though the issue occurred with the same error but it got resolved later in 2-3 minutes. I will look into it again to get from where the issue is originating.
```
error during apply of object "autoscaling.k8s.io/v1beta2/VerticalPodAutoscaler/kube-system/cloud-node-manager":
      no matches for kind "VerticalPodAutoscaler" in version "autoscaling.k8s.io/v1beta2"'
```

**Comment by rfranzke** [comment]
> Also, I don't understand why the error message shows autoscaling.k8s.io/v1beta2 haven't we already moved to autoscaling.k8s.io/v1.

Extensions which have not yet been released or deployed with new version might still try to apply `autoscaling.k8s.io/v1beta2.VPA` resources.

**Comment by ialidzhikov** [comment]
> I checked once again it's not always comparing by name. It first tries to sort the object based on kind but if kind is not there in the list which we provide then it sorts based on the string comparison of a kind and if kind is the same then it sorts based on the object name.
> Previously, VPA and others were not so they were getting deployed in last.

You can try to reconcile the following MR against https://github.com/gardener/gardener/pull/5822:

```yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: managedresource-example1
  namespace: default
type: Opaque
stringData:
  objects.yaml: |
    apiVersion: "autoscaling.k8s.io/v1"
    kind: VerticalPodAutoscaler
    metadata:
      name: cloud-node-manager
      namespace: kube-system
    spec:
      resourcePolicy:
        containerPolicies:
        - containerName: cloud-node-manager
          controlledValues: RequestsOnly
          minAllowed:
            cpu: 20m
            memory: 25Mi
      targetRef:
        apiVersion: apps/v1
        kind: DaemonSet
        name: cloud-node-manager
      updatePolicy:
        updateMode: "Auto"
    ---
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRole
    metadata:
      name: system:cloud-controller-manager
      annotations:
        rbac.authorization.kubernetes.io/autoupdate: "true"
      labels:
        k8s-app: cloud-controller-manager
    rules:
      - apiGroups:
          - ""
        resources:
          - events
        verbs:
          - create
          - patch
          - update
    ---
    kind: ClusterRoleBinding
    apiVersion: rbac.authorization.k8s.io/v1
    metadata:
      name: system:cloud-controller-manager
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: ClusterRole
      name: system:cloud-controller-manager
    subjects:
    - kind: User
      name: system:cloud-controller-manager
    - kind: User
      name: cloud-controller-manager
---
apiVersion: resources.gardener.cloud/v1alpha1
kind: ManagedResource
metadata:
  name: example
  namespace: default
spec:
  secretRefs:
  - name: managedresource-example1
```

I added some dummy logging with `fmt.Println`:

```diff
diff --git a/pkg/resourcemanager/controller/managedresource/sort.go b/pkg/resourcemanager/controller/managedresource/sort.go
index 4d8a2f045..9ec9efc9c 100644
--- a/pkg/resourcemanager/controller/managedresource/sort.go
+++ b/pkg/resourcemanager/controller/managedresource/sort.go
@@ -15,6 +15,7 @@
 package managedresource

 import (
+       "fmt"
        "sort"

        resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
@@ -87,10 +88,13 @@ func (k *kindSorter) Less(i, j int) bool {
        second, bok := k.ordering[b.oldInformation.Kind]

        if !aok && !bok {
+               fmt.Printf("b.oldInformation.Kind = %+v\n", a.oldInformation.Kind)
+               fmt.Printf("b.oldInformation.Kind = %+v\n", b.oldInformation.Kind)
                // if both are unknown then sort alphabetically by kind and name
                if a.oldInformation.Kind != b.oldInformation.Kind {
                        return a.oldInformation.Kind < b.oldInformation.Kind
                }
+               fmt.Println("Comparing by name")
                return a.oldInformation.Name < b.oldInformation.Name
        }
```

The result is:
```
{"level":"info","ts":"2022-06-02T12:36:07.681+0300","logger":"controller.resource-controller","msg":"Starting to reconcile ManagedResource","name":"example","namespace":"default"}

b.oldInformation.Kind =
b.oldInformation.Kind =
Comparing by name
b.oldInformation.Kind =
b.oldInformation.Kind =
Comparing by name
{"level":"info","ts":"2022-06-02T12:36:07.715+0300","logger":"resource-controller","msg":"Applying","resource":"autoscaling.k8s.io/v1/VerticalPodAutoscaler/kube-system/cloud-node-manager"}
{"level":"error","ts":"2022-06-02T12:36:07.725+0300","logger":"controller.resource-controller","msg":"Reconciler error","name":"example","namespace":"default","error":"could not apply all new resources: error during apply of object \"autoscaling.k8s.io/v1/VerticalPodAutoscaler/kube-system/cloud-node-manager\": no matches for kind \"VerticalPodAutoscaler\" in version \"autoscaling.k8s.io/v1\"","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/me/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/Users/me/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227"}
```

To me (as I mentioned in https://github.com/gardener/gardener/issues/5585#issuecomment-1118504930) kinds for some reason are empty and it is always comparing by name which leads the VPA to be applied before the ClusterRole and ClusterRoleBinding (`cloud-node-manager` comes before `system:cloud-controller-manager`).

The expected result is:
ClusterRole and ClusterRoleBinding to be applied first before the VPA.

I hope this helps you reproducing the issue.
